jax/tree_util.py

```diff
@@ -152,7 +160,15 @@ def tree_unflatten(cls, aux_data, children):
   register_pytree_node(cls, op.methodcaller('tree_flatten'), cls.tree_unflatten)
   return cls
 
-def tree_map(f, tree):
+@overload
+def tree_map(f: Callable[[U], U], tree: T) -> T:
+    ...
+
+@overload
+def tree_map(f: Callable[[T], U], tree: T) -> U:
```

Comment:
I still don't think this is a correct annotation. Counterexample:
```
def f(x: int) -> float: return x + 2.5
jax.tree_util.tree_map(f, {"foo": 3})
```
This cannot match the first overload because the input and output of `f` do not have the same type, and it cannot match the second overload because the input and output of `f` do not match the input and output of the `tree_map`.

I'm not convinced there is any meaningful type you can give this function that improves on the `Any` type you have below.

Similarly below.

Reply:
Yeah, great point.   It would be really nice to be able to keep the overloads since the first one covers the really common case where `f` maps array to array, transforming the components of a pytree.  If I add a third overload using `Any` as you suggest, then we get:

```python
def f(x: int) -> str:
    return "hi"

def g(x: int) -> int:
    return x

reveal_type(tree_map(f, 3))  # Revealed type is 'builtins.str*'
reveal_type(tree_map(f, {'foo': 3}))  # Revealed type is 'Any'
reveal_type(tree_map(g, [3, 3, 3]))  # Revealed type is 'builtins.list[builtins.int]'
```

Does that miss any important cases?

Reply:
Just pinging on this since I think this is all that's left to get this change pushed in.